File: solver.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <math.h>
#include <time.h>
#include <regex.h>
#include <stdint.h>

#define BUFFER_SIZE 1024

int check_command(const char *command) {
    char buffer[BUFFER_SIZE];
    snprintf(buffer, sizeof(buffer), "command -v %s > /dev/null 2>&1", command);
    return system(buffer) == 0;
}

void generate_random_string(char *str, size_t length) {
    const char charset[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    for (size_t i = 0; i < length; i++) {
        str[i] = charset[rand() % (sizeof(charset) - 1)];
    }
    str[length] = '\0';
}

double calculate_difficulty_raw(double difficulty) {
    return round(exp(log(256) * (4 - log(difficulty) / log(256))));
}

int main(int argc, char *argv[]) {
    if (!check_command("argon2") || !check_command("xxd") || !check_command("bc")) {
        printf("Required utilities are not installed. Please install them and try again.\n");
        return 1;
    }

    char challenge[BUFFER_SIZE];
    if (argc < 2) {
        printf("Enter Challenge Code: ");
        fgets(challenge, sizeof(challenge), stdin);
    } else {
        strncpy(challenge, argv[1], sizeof(challenge) - 1);
    }
    challenge[strcspn(challenge, "\n")] = 0;

    regex_t regex;
    regcomp(&regex, "^([0-9]+):([0-9]+):([A-Za-z0-9]+):([0-9]+)$", REG_EXTENDED);
    regmatch_t matches[5];
    if (regexec(&regex, challenge, 5, matches, 0) != 0) {
        printf("Invalid challenge format. Expected format: memory_cost:time_cost:salt:difficulty\n");
        return 2;
    }

    char *memory_cost = strtok(challenge, ":");
    char *time_cost = strtok(NULL, ":");
    char *salt = strtok(NULL, ":");
    char *difficulty = strtok(NULL, ":");

    printf("Memory Cost: %s\n", memory_cost);
    printf("Time Cost: %s\n", time_cost);
    printf("Salt: %s\n", salt);
    printf("Difficulty: %s\n", difficulty);

    char pw_prefix[BUFFER_SIZE];
    snprintf(pw_prefix, sizeof(pw_prefix), "UNBLOCK-");
    generate_random_string(pw_prefix + strlen(pw_prefix), 8);
    strcat(pw_prefix, "-");

    double difficulty_raw = calculate_difficulty_raw(atof(difficulty));
    printf("Estimated iterations: %s\n", difficulty);
    printf("Time Cost: %s\n\n", time_cost);

    int n = 1;
    time_t start_time = time(NULL);

    while (1) {
        char pw[BUFFER_SIZE];
        snprintf(pw, sizeof(pw), "%s%d", pw_prefix, n);
        
        char command[BUFFER_SIZE];
        snprintf(command, sizeof(command), "echo -n \"%s\" | argon2 %s -t %s -k %s -p 1 -id -v 13 -r", pw, salt, time_cost, memory_cost);
        
        FILE *fp = popen(command, "r");
        if (fp == NULL) {
            printf("Failed to run command\n");
            exit(1);
        }
        
        char hash_bytes[9] = {0};
        fread(hash_bytes, 1, 8, fp);
        pclose(fp);

        uint32_t hash_value = (uint32_t)strtoul(hash_bytes, NULL, 16);
        if (hash_value < (uint32_t)difficulty_raw) {
            printf("\nSOLUTION FOUND\n");
            printf("Your unblock code is: %s\n", pw);
            printf("This is the code you enter into the site to pass the challenge.\n\n");
            return 0;
        }

        n++;
        printf("\rElapsed Time: %ld seconds.", (time(NULL) - start_time));
        fflush(stdout);
    }

    regfree(&regex);
    return 0;
}

File: solver.py
import subprocess
import sys
import time
import re
import os
import random
import math

# Check if required utilities are installed
required_utils = ['argon2', 'xxd', 'bc']

for utils in required_utils:
    if subprocess.call(f"command -v {utils}", shell=True) != 0:
            print(f"{utils} is not installed. Please install it and try again.")
                    sys.exit(1)

                    # Read the challenge code from argument or prompt user
                    if len(sys.argv) < 2:
                        challenge = input("Enter Challenge Code: ")
                        else:
                            challenge = sys.argv[1].strip()

                            # Validate the challenge format
                            if not re.match(r'^([0-9]+):([0-9]+):([A-Za-z0-9]+):([0-9]+)$', challenge):
                                print("Invalid challenge format. Expected format: memory_cost:time_cost:salt:difficulty")
                                    sys.exit(2)

                                    # Parse challenge code
                                    memory_cost, time_cost, salt, difficulty = challenge.split(':')
                                    memory_cost = memory_cost.strip()
                                    time_cost = time_cost.strip()
                                    salt = salt.strip()
                                    difficulty = difficulty.strip()

                                    # Debugging output
                                    print(f"Memory Cost: {memory_cost}")
                                    print(f"Time Cost: {time_cost}")
                                    print(f"Salt: {salt}")
                                    print(f"Difficulty: {difficulty}")

                                    # Generate prefix for the password
                                    pw_prefix = f"UNBLOCK-{''.join(random.choices('ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789', k=8))}-"
                                    difficulty_raw = int(math.exp(math.log(256) * (4 - math.log(int(difficulty), 256))))  # Calculate difficulty threshold

                                    print(f"Estimated iterations: {difficulty}")
                                    print(f"Time Cost: {time_cost}\n")

                                    n = 1
                                    start_time = time.time()

                                    # Main loop to find the solution
                                    while True:
                                        pw = f"{pw_prefix}{n}"
                                            # Create hash using argon2
                                                hash_result = subprocess.check_output(
                                                        f"echo -n '{pw}' | argon2 {salt} -t {time_cost} -k {memory_cost} -p 1 -id -v 13 -r",
                                                                shell=True, text=True, stderr=subprocess.DEVNULL
                                                                    )
                                                                        hash_bytes = hash_result.split('$')[3][:8]  # Extract the first 8 characters of the hash

                                                                            if int(hash_bytes, 16) < difficulty_raw:
                                                                                    print(f"\nSOLUTION FOUND")
                                                                                            print(f"Your unblock code is: {pw}")
                                                                                                    print("This is the code you enter into the site to pass the challenge.\n")
                                                                                                            break
                                                                                                                else:
                                                                                                                        elapsed_time = int(time.time() - start_time)
                                                                                                                                print(f"\rElapsed Time: {elapsed_time} seconds.", end="")
                                                                                                                                        n += 1